File: days/02/part1.py

def is_valid(game):
    """Checks the validity of the game"""

    for round in game.split('; '):
        for ball in round.split(', '):
            n, color = ball.split(' ')
            n = int(n)

            # If any ball count exceeds the limit, the game is invalid
            if n > limits[color]:
                return False

    return True

limits = {
    'red': 12,
    'green': 13,
    'blue': 14
}

with open('inputs/day02.txt') as file:
    total = 0
    for i,line in enumerate(file):
        _, game = line.split(': ')
        if is_valid(game.strip()):
            total += i+1

    print(total)

File: days/02/part1.rs
use std::io::{self, BufReader};
use std::fs::File;
use std::io::prelude::*;

fn main() -> io::Result<()> {
    let f = File::open("inputs/day02.txt")?;
    // Read file line by line
    let reader = BufReader::new(f);

    let mut sum = 0;
    for (i,line) in reader.lines().enumerate() {
        match line {
            Ok(l) => {
                let mut rounds = l.split(": ")
                    .nth(1)
                    .unwrap()
                    .split("; ");
                
                // Check if all rounds are valid
                let ok = rounds.all(|round| -> bool {

                    // Check if all numbers of balls are valid
                    round.split(", ").all(|ball| {
                        let s: Vec<&str> = ball.split(" ").take(2).collect();
                        let col = s[1];
                        let n = s[0].parse::<i32>().unwrap();
                        
                        match col {
                            "red" => if n > 12 { false } else {true},
                            "green" => if n > 13 { false } else {true},
                            "blue" => if n > 14 { false } else {true},
                            _ => false,
                        }

                    })
                });

                if ok {
                    sum += i+1;
                }

            },
            Err(e) => println!("{}", e),
        }
    }

    println!("{}", sum);

    Ok(())
}